PROTOCOL.md

```diff
@@ -169,36 +180,79 @@ A delta file, `n.json`, contains an atomic set of [_actions_](#Actions) that sho
 
 ### Checkpoints
-Checkpoints are also stored in the `_delta_log` directory, and can be created for any version of the table.
+Checkpoints are also stored in the `_delta_log` directory, and can be created at any time, for any committed version of the table.
+For performance reasons, readers should prefer to use the newest complete checkpoint possible.
+For time travel, the checkpoint used must not be newer than the time travel version.
 
-A checkpoint contains the complete replay of all actions up until this version, with invalid actions removed.
-Invalid actions are those that have been canceled out by a subsequent ones (for example removing a file that has been added), using the [rules for reconciliation](#Action-Reconciliation)
-Checkpoints allow readers to short-cut the cost of reading the log up-to a given point in order to reconstruct a snapshot, and allow older JSON Delta log entries to be deleted after a period of time.
+A checkpoint contains the complete replay of all actions, up to and including the checkpointed table version, with invalid actions removed.
+Invalid actions are those that have been canceled out by a subsequent ones (for example removing a file that has been added), using the [rules for reconciliation](#Action-Reconciliation).
+In addition to above, checkpoint also contains the [_remove tombstones_](#add-file-and-remove-file) until they are expired.
+Checkpoints allow readers to short-cut the cost of reading the log up-to a given point in order to reconstruct a snapshot, and they also allow [Metadata cleanup](#metadata-cleanup) to delete expired JSON Delta log entries.
 
-
-By default, the reference implementation creates a checkpoint every 10 commits.
+Readers SHOULD NOT make any assumptions about the existence or frequency of checkpoints, with one exception:
+[Metadata cleanup](#metadata-cleanup) MUST provide a checkpoint for the oldest kept table version, to cover all deleted [Delta log entries](#delta-log-entries).
+That said, writers are encouraged to checkpoint reasonably frequently, so that readers do not pay excessive log replay costs due to reading large numbers of delta files.
 
 The checkpoint file name is based on the version of the table that the checkpoint contains.
-The format of the checkpoint file name can take one of two forms:
 
-1. A single checkpoint file for version `n` of the table will be named `n.checkpoint.parquet`. For example:
+Delta supports three kinds of checkpoints:
+
+1. UUID-named Checkpoints: These follow [V2 spec](#v2-spec) which uses the following file name: `n.checkpoint.u.{json/parquet}`, where `u` is a UUID and `n` is the
+snapshot version that this checkpoint represents. The UUID-named V2 Checkpoint may be in json or parquet format, and references zero or more checkpoint sidecars
+in the `_delta_log/_sidecars` directory. A checkpoint sidecar is a uniquely-named parquet file: {unique}.parquet" where unique is some unique
```

Comment:
```suggestion
in the `_delta_log/_sidecars` directory. A checkpoint sidecar is a uniquely-named parquet file: `{unique}.parquet` where `unique` is some unique
```

Reply:
We are not putting restriction on length as of now - today we don't put constrain name on data file names which are arbitrary parquet files. Following same model here.